sql-plugin/src/main/scala/com/nvidia/spark/rapids/AllocationRetryCoverageTracker.scala

```diff
@@ -0,0 +1,230 @@
+ * 
+ * This feature is DISABLED by default. To enable, set environment variable:
+ *   export SPARK_RAPIDS_RETRY_COVERAGE_TRACKING=true
+ * 
+ * For integration tests:
+ *   SPARK_RAPIDS_RETRY_COVERAGE_TRACKING=true ./integration_tests/run_pyspark_from_build.sh
+ * 
+ * When enabled:
+ * - Checks memory allocations' call stacks for retry-related methods
+ * - Logs uncovered allocations to /tmp/uncovered_allocations.csv
+ * - Also logs warnings via Spark logging
+ * 
+ * Note: Enabling this feature turns on RMM debug mode which may impact performance.
+ * Only use for debugging/testing purposes.
+ * 
+ * See: https://github.com/NVIDIA/spark-rapids/issues/13672
+ */
+object AllocationRetryCoverageTracker extends Logging {
+  import AllocationKind._
+
+  // Package prefixes for spark-rapids code where we look for "Retry" in method names
+  private val SPARK_RAPIDS_PACKAGE_PREFIXES = Seq(
+    "com.nvidia.spark.rapids",
+    "org.apache.spark.sql.rapids"
+  )
+
+  // Default output file path for uncovered allocations - use /tmp for easy access
+  private val DEFAULT_OUTPUT_PATH = "/tmp/uncovered_allocations.csv"
+
+  // Environment variable to enable tracking
+  private val ENV_VAR_NAME = "SPARK_RAPIDS_RETRY_COVERAGE_TRACKING"
+
+  // Check environment variable - this works reliably across all processes
+  val ENABLED: Boolean = {
+    val envValue = System.getenv(ENV_VAR_NAME)
+    envValue != null && envValue.equalsIgnoreCase("true")
+  }
+
+  @volatile private var headerWritten: Boolean = false
+
+  // Track unique call stacks we've already logged to avoid duplicates
+  private val loggedStacks = ConcurrentHashMap.newKeySet[String]()
+
+  // Lock for thread-safe file writing
+  private val writeLock = new Object()
+
+  /**
+   * Ensure header is written (thread-safe, lazy initialization).
+   */
+  private def ensureHeaderWritten(): Unit = {
+    if (!headerWritten) {
+      writeLock.synchronized {
+        if (!headerWritten) {
+          logWarning(s"Retry coverage tracking ACTIVE. Output: $DEFAULT_OUTPUT_PATH")
+          writeToFileInternal("kind,call_stack", append = false)
+          headerWritten = true
+        }
+      }
+    }
+  }
+
+  /**
+   * Check if a memory allocation is covered by a retry method.
+   * If not covered and tracking is enabled, log to the output file.
+   * 
+   * @param kind The kind of memory allocation (HOST or DEVICE)
+   */
+  private def checkAllocationInternal(kind: AllocationKind): Unit = {
+    // Ensure header is written on first check
+    ensureHeaderWritten()
+
+    val stackTrace = Thread.currentThread().getStackTrace
+    
+    // Check if any retry-related code is in the call stack.
+    // We look for any class or method containing "Retry" in its name within spark-rapids packages.
+    // This catches:
+    // - Core retry methods: withRetry, withRetryNoSplit, withRestoreOnRetry
+    // - Wrapper functions: concatenateAndMergeWithRetry, projectAndCloseWithRetrySingleBatch, etc.
```

Comment:
Not sure if these are duplicate, since they already have `withRetryNoSplit` call inside.